qiime2/sdk/usage.py

```diff
@@ -159,80 +431,160 @@ def build_opts(self, action_signature, scope):
 
-    @property
-    def records(self):
-        return types.MappingProxyType(self._records)
+        Parameters
+        ----------
+        inputs : UsageInputs
+        outputs : UsageOutputNames
 
-    def push_record(self, ref, value, source, assert_has_line_matching=None):
-        record = ScopeRecord(ref=ref, value=value, source=source,
-                             assert_has_line_matching=assert_has_line_matching)
-        self._records[ref] = record
-        return record
+        """
+        if not isinstance(inputs, UsageInputs):
+            raise TypeError('Must provide an instance of UsageInputs.')
+        if not isinstance(outputs, UsageOutputNames):
+            raise TypeError('Must provide an instance of UsageOutputNames.')
 
-    def get_record(self, ref):
-        try:
-            return self.records[ref]
-        except KeyError:
-            raise KeyError('No record with ref id: "%s" in scope.' % (ref, ))
+        _, sig = self.get_action()
+
+        inputs.validate(sig)
+        outputs.validate(sig)
 
 
 class Usage(metaclass=abc.ABCMeta):
+    """`Usage` is the base class for Usage driver implementations.
+    """
+
     def __init__(self):
         self._scope = Scope()
 
-    def init_data(self, ref, factory):
+    def init_data(
+        self, ref: str, factory: typing.Callable[[], "sdk.Artifact"]
+    ) -> ScopeRecord:
+        """Initialize example data from a factory.
+
+        Parameters
+        ----------
+        ref : str
+            Unique name for example data
+        factory : callable
+            A factory that returns an example Artifact
+
+        Returns
+        -------
+        record : ScopeRecord
+            A record with information about example data
+        """
         value = self._init_data_(ref, factory)
         return self._push_record(ref, value, 'init_data')
 
     def _init_data_(self, ref, factory):
         raise NotImplementedError
 
-    def init_metadata(self, ref, factory):
+    def init_metadata(
+        self, ref: str, factory: typing.Callable[[], "metadata.Metadata"]
+    ) -> ScopeRecord:
+        """Initialize metadata for a Usage example.
+
+        Parameters
+        ----------
+        ref : str
+            Unique name for example metadata
+        factory : callable
+            A factory that returns example Metadata
+
+        Returns
+        -------
+        record : ScopeRecord
+            A record with information about example Metadata
+        """
         value = self._init_metadata_(ref, factory)
         return self._push_record(ref, value, 'init_metadata')
 
     def _init_metadata_(self, ref, factory):
         raise NotImplementedError
 
-    def init_data_collection(self, ref, collection_type, *records):
+    def init_data_collection(
+        self,
+        ref: str,
+        collection_type: typing.Union[list, set],
+        *records: ScopeRecord
+    ) -> ScopeRecord:
+        """Initialize a collection of data for a Usage example.
```

Comment:
```suggestion
    def init_data_collection(self, ref: str,
                             collection_type: typing.Union[list, set],
                             *records: typing.List[ScopeRecord]) \
                             -> ScopeRecord:
        """
        Initialize a collection of data for a Usage example.
```

Reply:
I think `*records: ScopeRecord` is the correct type hint here (see https://www.python.org/dev/peps/pep-0484/#arbitrary-argument-lists-and-default-argument-values).

Reply:
Thanks, I agree with your finding!